commit 78ca77e933b9a2d809e4151edc90821a32803c1f
Author: Franco Rapetti <frapeti@gmail.com>
Date:   Tue Nov 26 16:36:53 2013 -0200

    Port incall volume workaround to cm11 - q
    Conflicts:
    	src/com/android/phone/PhoneUtils.java
    
    Change-Id: Ib7789137e8d669d08dfc6edc44463e384e752e8a

diff --git a/src/com/android/phone/PhoneUtils.java b/src/com/android/phone/PhoneUtils.java
index 2d44977..f1b3350 100644
--- a/src/com/android/phone/PhoneUtils.java
+++ b/src/com/android/phone/PhoneUtils.java
@@ -2765,4 +2765,35 @@ public class PhoneUtils {
         return context.getResources().getConfiguration().orientation
                 == Configuration.ORIENTATION_LANDSCAPE;
     }
+
+     /**
+     * Reset the audio stream volume to fix the low in-call volume bug.
+     *
+     * Due to a bug in the OMX system, the audio stream volume is set to 0 after it was set to it's default volume.
+     * Calling PhoneUtils.resetAudioStreamVolume() triggers the system to reset the volume.
+     *
+     * This should be called on every place where is switched between audio modes.
+     *
+     * REMARK: I think it only appears on the voice call stream, but to be sure I also do it on the bluetooth stream.
+     */
+    static void resetAudioStreamVolume() {
+        PhoneGlobals app = PhoneGlobals.getInstance();
+        AudioManager audioManager = (AudioManager) app.getSystemService(Context.AUDIO_SERVICE);
+        // determine actual streamType
+        int streamType = AudioManager.STREAM_VOICE_CALL;
+        if (app.isBluetoothHeadsetAudioOn()) {
+            streamType = AudioManager.STREAM_BLUETOOTH_SCO;
+        }
+        // determine volume and 1 level lower volume (lowest level can be 0)
+        int volume = audioManager.getStreamVolume(streamType);
+        int lowerVolume = volume - 1;
+        if (lowerVolume < 0) {
+            lowerVolume = 0;
+        }
+        log("resetAudioStreamVolume (streamType=" + streamType + ", streamVolume=" + volume + ")...");
+        // It's important to change it to another volume before restoring the original volume,
+        // otherwise the volume change will NOT be triggered!!
+        audioManager.setStreamVolume(streamType, lowerVolume, 0);
+        audioManager.setStreamVolume(streamType, volume, 0);
+    }
 }

commit e60bfb6ccb7ff29fceb6129c88d4e83d7d1fa9ed
Author: Franco Rapetti <frapeti@gmail.com>
Date:   Tue Nov 26 16:59:21 2013 -0200

    reset the audio volume stream after switching audio mode

diff --git a/src/com/android/phone/AudioRouter.java b/src/com/android/phone/AudioRouter.java
index 3c8e9d3..1d136a2 100644
--- a/src/com/android/phone/AudioRouter.java
+++ b/src/com/android/phone/AudioRouter.java
@@ -350,6 +350,10 @@ import java.util.List;
         if (doNotify) {
             notifyListeners();
         }
+
+        // Fix for low in-call volume bug.
+        // Reset the audio volume stream after switching audio mode
+        PhoneUtils.resetAudioStreamVolume();
     }
 
     /**

commit 490d50730b7217c2c26af06db8b67b6d31933346
Author: Franco Rapetti <frapeti@gmail.com>
Date:   Tue Nov 26 21:32:48 2013 -0200

    get BluetoothManager to ask about bt headset (KK implementation)

diff --git a/src/com/android/phone/PhoneUtils.java b/src/com/android/phone/PhoneUtils.java
index f1b3350..5bb8bc8 100644
--- a/src/com/android/phone/PhoneUtils.java
+++ b/src/com/android/phone/PhoneUtils.java
@@ -2780,8 +2780,11 @@ public class PhoneUtils {
         PhoneGlobals app = PhoneGlobals.getInstance();
         AudioManager audioManager = (AudioManager) app.getSystemService(Context.AUDIO_SERVICE);
         // determine actual streamType
+
+        final BluetoothManager btManager = app.getBluetoothManager();
+
         int streamType = AudioManager.STREAM_VOICE_CALL;
-        if (app.isBluetoothHeadsetAudioOn()) {
+        if (btManager.isBluetoothHeadsetAudioOn()) {
             streamType = AudioManager.STREAM_BLUETOOTH_SCO;
         }
         // determine volume and 1 level lower volume (lowest level can be 0)
